# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

Remember!
- use this to get data from csv file- open terminal - sqlite3 Movies.db - follow "here" link
- .table will give list of tables- can check for duplicates!!
- PRAGMA will give metadata

In [11]:
%%sql

SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(DATASET);

 * sqlite:///Movies.db
Done.
Done.


[(0, 'TName', 'TEXT', 0, None, 0),
 (1, 'Location', 'TEXT', 0, None, 0),
 (2, 'Phone', 'TEXT', 0, None, 0),
 (3, 'MTitle', 'TEXT', 0, None, 0),
 (4, 'ShowTime', 'TEXT', 0, None, 0),
 (5, 'Rating', 'TEXT', 0, None, 0),
 (6, 'CCode', 'TEXT', 0, None, 0),
 (7, 'CName', 'TEXT', 0, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [8]:
%%sql

DELETE FROM MOVIES;
SELECT DISTINCT MTitle, Rating FROM DATASET;

 * sqlite:///Movies.db
0 rows affected.
Done.


[('Associate, The', 'PG-13'),
 ('Ghost & The Darkness, The', 'R'),
 ('Independence Day', 'PG-13'),
 ('D3: The Mighty Ducks', 'PG'),
 ('Dear God', ''),
 ('First Wives Club, The', 'PG-13'),
 ('High School High', 'PG-13'),
 ('Larger Than Life', 'PG'),
 ('Mirror Has Two Faces, The', 'PG-13'),
 ('Ransom', 'R'),
 ('Sleepers', ''),
 ('Space Jam', 'PG'),
 ("Stephen King's Thinner", ''),
 ('Michael Collins', 'R'),
 ('Romeo & Juliet', 'R'),
 ('Big Night', 'R'),
 ('Fly Away Home', 'PG'),
 ('Nutty Professor, The', 'PG-13'),
 ('Phenomenon', 'PG'),
 ('Set It Off', 'R'),
 ('Surviving Picasso', 'R'),
 ('Very Brady Sequel, A', 'PG-13'),
 ('That Thing You Do', 'PG')]

lets insert into the table!
REMEMBER: INSERT INTO table (part 2 titles)
          SELECT DISTINCT (columns from DATASET)
Shows- joining the other two tables to retrieve data for the shows table

In [29]:
%%sql

DELETE FROM Movies;
INSERT INTO Movies (Title, Rating)
    SELECT DISTINCT MTitle, Rating FROM DATASET;
SELECT * FROM MOVIES;


DELETE FROM Theaters;
INSERT INTO Theaters (Name, Location, Phone)
    SELECT DISTINCT TName, Location, Phone FROM DATASET;
SELECT * FROM Theaters;


DELETE FROM Artists;
INSERT INTO Artists (Name)
    SELECT DISTINCT CName FROM DATASET;
SELECT * FROM Artists;


DELETE FROM Shows;
INSERT INTO Shows (ShowTime, MID, TID)
    SELECT DISTINCT DATASET.ShowTime, MOVIES.MID, THEATERS.TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.Title = DATASET.MTitle)
        JOIN THEATERS ON (THEATERS.Name = DATASET.TName);
SELECT ShowTime, Title, Name FROM Shows JOIN Movies USING (MID) JOIN Theaters USING (TID);

DELETE FROM Credit;
INSERT INTO Credit (CCode, AID, MID)
    SELECT DISTINCT DATASET.CCode, MOVIES.MID, ARTISTS.AID
    FROM DATASET
        JOIN ARTISTS ON (ARTISTS.NAME = DATASET.CName)
        JOIN MOVIES ON (MOVIES.Title = DATASET.MTitle);
SELECT * FROM CREDIT LIMIT 10;

 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
Done.
9 rows affected.
9 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
131 rows affected.
131 rows affected.
Done.
162 rows affected.
162 rows affected.
Done.


[(1, 1, 1, 'A'),
 (2, 1, 2, 'A'),
 (3, 1, 3, 'A'),
 (4, 1, 4, 'A'),
 (5, 1, 5, 'A'),
 (6, 1, 6, 'A'),
 (7, 1, 7, 'A'),
 (8, 1, 8, 'A'),
 (9, 1, 9, 'D'),
 (10, 2, 10, 'A')]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [ ]:
YOUR CODE HERE

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".